### PR TITLE
android: release AudioTrack at Close()

### PR DESCRIPTION
**Pull request by pankona**
Here's Pull Request for issue #27 .

* Since there's a limit to the number of AudioTrack that can exist on same time, allocated AudioTrack should be released after its use.
* I think mutex lock/unlock is necessary for Close() because release method for "released AudioTrack" throws  IllegalStateException

### CODE REVIEW

**Comment by pankona** [comment]
@hajimehoshi 
Sorry to be late. Here's pull request for the issue.
I'm happy if you check the modification. I hope this will help!

File: player_android.go [hunk 1]
```diff
@@ -279,6 +292,19 @@ func (p *player) Write(data []uint8) (int, error) {
 }
 
 func (p *player) Close() error {
+	p.mu.Lock()
+	p.mu.Unlock()
```
Comment (hajimehoshi) [inline player_android.go]: What is this lock?

Comment (pankona) [inline player_android.go]: It seems IllegalStateException is throwed if release method is called for the AudioTrack instance that is already released.
I inserted these mutex lock/unlock and check nil of `p.audioTrack` to care for concurrent call of `Close()` function.

Comment (pankona) [inline player_android.go]: ah, sorry, I need to insert `defer` ...

Comment (hajimehoshi) [inline player_android.go]: > It seems IllegalStateException is throwed if release method is called for the AudioTrack instance that is already released.

Hm? So `Close` callers should have responsibility of concurrency, and `Close` doesn't have to care about this, right?

> ah, sorry, I need to insert defer ...
  
Yes, thanks. But in the first place, I'm wondering if we really need to lock here.

Comment (pankona) [inline player_android.go]: > I think Close callers should have responsibility of concurrency,

Ah, understood. Sorry for misunderstanding of whole policy and design.
Let me update again not to use mutexes.

Comment (pankona) [inline player_android.go]: Updated (with ammend). new commit is 1e78c26 .
Mutexes are removed. Would you please review them again?
Thank you so much for your cooperation!

File: player_android.go [hunk 2]
```diff
@@ -279,6 +290,19 @@ func (p *player) Write(data []uint8) (int, error) {
...
 
 func (p *player) Close() error {
+	if p.audioTrack == nil {
+		return nil
+	}
+
 	runtime.SetFinalizer(p, nil)
-	return nil
+	err := jni.RunOnJVM(func(vm, env, ctx uintptr) error {
+		msg := C.releaseAudioTrack(C.uintptr_t(vm), C.uintptr_t(env), p.audioTrack)
```
Comment (hajimehoshi) [inline player_android.go]: I prefer one line check `if msg := ...; msg != nil { ... }`

Comment (pankona) [inline player_android.go]: Sure thing, I will!

Comment (pankona) [inline player_android.go]: Updated. 5682ff6
Refer to the source codes around here, 
https://github.com/hajimehoshi/oto/blob/8ab1cc50f21ad05f93bf99dc8c44d6aa101c29c9/player_android.go#L209-L213
I inserted new line in condition on L.299-L300 too, do you prefer it?

Comment (hajimehoshi) [inline player_android.go]: I don't care (both are fine). Thank you, lgtm

**Comment by pankona** [comment]
Thank you so much for your review and merge! :smile:

**Pull request merged**